demo/kotlin/src/main/kotlin/LunarEclipse.kt

```diff
@@ -7,16 +7,11 @@ import io.github.cosinekitty.astronomy.*
  * The date and time after which to start searching for lunar eclipses.
  */
 internal fun `Lunar eclipse demo`(startTime: Time): Int {
-    var e: LunarEclipseInfo = searchLunarEclipse(startTime)
-    var count = 0
-    while (true) {
-        if (e.kind != EclipseKind.Penumbral) {
-            printEclipse(e)
-            if (++count == 10)
-                return 0
-        }
-        e = nextLunarEclipse(e.peak)
-    }
+    generateSequence(searchLunarEclipse(startTime)) { nextLunarEclipse(it.peak) }
```

Comment:
To follow up about generic iterators, I suppose in Kotlin all I need is something roughly like

```kotlin
fun lunarEclipsesAfter(startTime: Time) = 
    generateSequence(searchLunarEclipse(startTime)) { nextLunarEclipse(it.peak) }
```

Reply:
Makes sense